File: ch1/1-24.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
)

func main() {
	r := bufio.NewReader(os.Stdin)
	for {
		line, err := r.ReadBytes(byte('\n'))
		if err != nil {
			if err != io.EOF {
				fmt.Printf("error: %s\n", err)
			}
			break
		}
		l := len(line)
		var commErr bool
		var (
			parensL   int
			bracketsL int
			doubleQl  int
			singleQl  int
		)
		for i := 0; i < l; i++ {
			ch := rune(line[i])
			if ch == '(' || ch == ')' {
				parensL++
			}
			if ch == '[' || ch == ']' {
				bracketsL++
			}
			if ch == '"' {
				doubleQl++
			}
			if ch == '\'' {
				singleQl++
			}
			if ch == '/' && rune(line[i+1]) == '/' {
				break
			} else if ch == '/' {
				commErr = true
			}
		}
		if parensL%2 != 0 || bracketsL%2 != 0 || singleQl%2 != 0 || doubleQl%2 != 0 || commErr == true {
			fmt.Println("syntax error!")
		}
	}
}

File: ch1/1-24.c
#include <stdio.h>

#define MAXLINE 1000

int get_line(char s[], int lim);

int main(void)
{
    char line[MAXLINE];
    int len;
    while ((len = get_line(line, MAXLINE)) > 0) {
        int parens_l = 0;
        int brackets_l = 0;
        int double_q_l = 0;
        int single_q_l = 0;
        int comm_err = 0;
        for (int i = 0; i < len; i++) {
            if (line[i] == '(' || line[i] == ')') {
                parens_l++;
            }
            if (line[i] == '[' || line[i] == ']') {
                brackets_l++;
            }
            if (line[i] == '"') {
                double_q_l++;
            }
            if (line[i] == '\'') {
                single_q_l++;
            }
            if (line[i] == '/' && line[i+1] == '/') {
                break;
            } else if (line[i] == '/') {
                comm_err = 1;
            }
        }
        if (parens_l%2 != 0 || brackets_l%2 != 0 || single_q_l%2 != 0 || double_q_l%2 != 0 || comm_err != 0) {
            puts("Syntax error!");
        }
    }
    return 0;
}

int get_line(char s[], int lim)
{
    int c, i;
    for (i = 0; i < lim - 1 && (c = getchar()) != EOF && c != '\n'; i++) {
        s[i] = c;
    }
    if (c == '\n') {
        s[i] = c;
        ++i;
    }
    s[i] = '\0';

    return i;
}
